src/js/legend.js

```diff
@@ -0,0 +1,228 @@
+                    args: ["{that}.options.legendOptions.colors"]
+                }
+            }
+        },
+        modelRelay: [{
+            target: "dataSetWithColors",
+            singleTransform: {
+                type: "fluid.transforms.free",
+                args: {
+                    "dataSet": "{that}.model.dataSet",
+                    "color": "{that}.color"
+                },
+                func: "gpii.chartAuthoring.pieChart.legend.consolidateDataAndColors"
+            }
+        }],
+        styles: {
+            legend: "gpii-ca-pieChart-legend",
+            table: "gpii-ca-pieChart-legend-table",
+            row: "gpii-ca-pieChart-legend-table-row",
+            legendColorCell: "gpii-ca-pieChart-legend-color-cell",
+            labelCell: "gpii-ca-pieChart-legend-label-cell",
+            valueCell: "gpii-ca-pieChart-legend-value-cell"
+        },
+        selectors: {
+            legend: ".gpiic-ca-pieChart-legend",
+            table: ".gpii-ca-pieChart-legend-table",
+            row: ".gpii-ca-pieChart-legend-table-row",
+            legendColorCell: ".gpii-ca-pieChart-legend-color-cell",
+            labelCell: ".gpii-ca-pieChart-legend-label-cell",
+            valueCell: ".gpii-ca-pieChart-legend-value-cell"
+        },
+        events: {
+            onLegendCreated: null  // Fire when the legend is created. Ready to register D3 DOM event listeners
+        },
+        listeners: {
+            "onCreate.create": {
+                funcName: "gpii.chartAuthoring.pieChart.legend.create",
+                args: ["{that}"]
+            }
+        },
+        modelListeners: {
+            dataSet: {
+                funcName: "{that}.draw",
+                excludeSource: "init"
+            }
+        },
+        invokers: {
+            draw: {
+                funcName: "gpii.chartAuthoring.pieChart.legend.draw",
+                args: ["{that}"]
+            }
+        }
+    });
+
+    // Return a D3 color scale based on user supplied colors or the d3.scale.category10() defaults
+
+    gpii.chartAuthoring.pieChart.legend.getColorScale = function (colors) {
+        return colors ? d3.scale.ordinal().range(colors) : d3.scale.category10();
+    };
+
+    // Takes the dataSet array and the color array, and returns a consolidated object array to ease sorting and other operations while keeping colors "correct"
+
+    gpii.chartAuthoring.pieChart.legend.consolidateDataAndColors = function (model) {
+        var dataSet = model.dataSet;
+        var color = model.color;
+
+        var c = [];
+        fluid.each(dataSet, function(item, index) {
+            var d = {
+                id: item.id,
+                label: item.label,
+                value: item.value,
+                color: color(index)
+            };
+            c.push(d);
+        });
+
+        return c;
+
+    };
+
+    gpii.chartAuthoring.pieChart.legend.draw = function (that) {
```

Comment:
This function contains within it many function definitions. We should break these out into reusable functions wherever possible.